sycl/source/builtins_generator.py

```diff
@@ -0,0 +1,1285 @@
+  def __init__(self, related_arg_type, templated_type):
+    self.related_arg_type = related_arg_type
+    self.templated_type = templated_type
+
+  def __str__(self):
+    if isinstance(self.templated_type, Vec):
+      # Vectors need conversion as their related type may be a swizzle.
+      return f'detail::get_vec_t<{self.related_arg_type}>'
+    return str(self.related_arg_type)
+
+# A class representing an instantiated version of ElementType containing the
+# referenced type.
+# referenced_type - The referenced type.
+# parent_idx - The index of the parent argument in the corresponding builtin.
+class InstantiatedElementType:
+  def __init__(self, referenced_type, parent_idx):
+    self.referenced_type = referenced_type
+    self.parent_idx = parent_idx
+
+  def __str__(self):
+    if (isinstance(self.referenced_type, InstantiatedTemplatedArg) or
+        isinstance(self.referenced_type, InstantiatedTemplatedReturnType)):
+      return f'detail::get_elem_type_t<{self.referenced_type}>'
+    return self.referenced_type
+
+# A class representing an instantiated version of ConversionTraitType containing
+# the parent type.
+# parent_type - The parent type.
+# trait - The name of the trait to use for converting the type.
+# parent_idx - The index of the parent argument in the corresponding builtin.
+class InstantiatedConversionTraitType:
+  def __init__(self, parent_type, trait, parent_idx):
+    self.parent_type = parent_type
+    self.trait = trait
+    self.parent_idx = parent_idx
+
+  def __str__(self):
+    return f'detail::{self.trait}<{self.parent_type}>'
+
+### GENTYPE DEFINITIONS
+
+vfloatn = [Vec(["float"])]
+vfloat3or4 = [Vec(["float"], {3,4})]
+mfloatn = [Marray(["float"])]
+mfloat3or4 = [Marray(["float"], {3,4})]
+genfloatf = ["float", Vec(["float"]), Marray(["float"])]
+
+vdoublen = [Vec(["double"])]
+vdouble3or4 = [Vec(["double"], {3,4})]
+mdoublen = [Marray(["double"])]
+mdouble3or4 = [Marray(["double"], {3,4})]
+
+vhalfn = [Vec(["half"])]
+vhalf3or4 = [Vec(["half"], {3,4})]
+mhalfn = [Marray(["half"])]
+mhalf3or4 = [Marray(["half"], {3,4})]
+
+genfloat = ["float", "double", "half", Vec(["float", "double", "half"]),
+            Marray(["float", "double", "half"])]
+vgenfloat = [Vec(["float", "double", "half"])]
+sgenfloat = ["float", "double", "half"]
+mgenfloat = [Marray(["float", "double", "half"])]
+
+vgeofloat = [Vec(["float"], {1,2,3,4})]
+vgeodouble = [Vec(["double"], {1,2,3,4})]
+vgeohalf = [Vec(["half"], {1,2,3,4})]
+mgeofloat = [Marray(["float"], {1,2,3,4})]
+mgeodouble = [Marray(["double"], {1,2,3,4})]
+mgeohalf = [Marray(["half"], {1,2,3,4})]
+gengeofloat = ["float", Vec(["float"], {1,2,3,4}), Marray(["float"], {1,2,3,4})]
+gengeodouble = ["double", Vec(["double"], {1,2,3,4}), Marray(["double"], {1,2,3,4})]
+gengeohalf = ["half", Vec(["half"], {1,2,3,4}), Marray(["half"], {1,2,3,4})]
+
+vint8n = [Vec(["int8_t"])]
+vint16n = [Vec(["int16_t"])]
+vint32n = [Vec(["int32_t"])]
+vint64n = [Vec(["int64_t", "long long"])]
```

Comment:
Why `long long` added as an alias here? It is not strictly 64-bit type, it only guaranteed to be **at least** 64-bit: [cppreference](https://en.cppreference.com/w/cpp/language/types)

Reply:
Right! However, we previously had vector definitions with this type in our builtin definitions and so to avoid breaking user code without deprecation, I added it here to. It is not part of the spec.

Note also that I may have to add more cases like this. Right now they are not deprecated here, but I want to make a follow-up patch deprecating the non-standard ones, although to do so we need to do some extra checks to avoid creating conflicts when the fixed-size integers are simply aliasing the fundamental types.